datastore/providers/postgres_datastore.py

```diff
@@ -0,0 +1,211 @@
+            % (POSTGRES_TABLENAME, str(VECTOR_DIMENSION))
+        )
+        cur = conn.cursor()
+        cur.execute(create_table)
+        # Commit the transaction
+        conn.commit()
+        # Close the cursor and the connection
+        cur.close()
+        conn.close()
+
+    async def _upsert(self, chunks: Dict[str, List[DocumentChunk]]) -> List[str]:
+        """
+        Takes in a dict from document id to list of document chunks and inserts them into the index.
+        Return a list of document ids.
+        """
+        # Create a connection
+        conn = psycopg.connect(**self.conn_params)
+        cur = conn.cursor()
+
+        # Initialize a list of ids to return
+        doc_ids: List[str] = []
+
+        # Set synchronous commit
+        cur.execute("SET synchronous_commit = %s" % POSTGRES_SYNCHRONOUS_COMMIT)
+
+        with cur.copy(
+            "COPY %s (doc_id, chunk_id, text, embedding, metadata) FROM STDIN"
+            % POSTGRES_TABLENAME
+        ) as copy:
+            for doc_id, chunk_list in chunks.items():
+                doc_ids.append(doc_id)
+                for chunk in chunk_list:
+                    text = sql.Literal(chunk.text).as_string(conn)
+                    embedding = "[" + ",".join([str(v) for v in chunk.embedding]) + "]"
+
+                    metadata = chunk.metadata.dict()
+                    if "created_at" in list(metadata.keys()):
+                        if metadata["created_at"]:
+                            metadata["created_at"] = to_unix_timestamp(
+                                metadata["created_at"]
+                            )
+                    metadata = json.dumps(metadata)
+
+                    row = (doc_id, chunk.id, text, embedding, metadata)
+                    copy.write_row(row)
+
+        index_statement = (
+            "CREATE INDEX ON %s USING ivfflat (embedding vector_cosine_ops)"
+            % POSTGRES_TABLENAME
+        )
+        cur.execute(index_statement)
+
+        conn.commit()
+        cur.close()
+        conn.close()
+
+        return doc_ids
+
+    async def _query(self, queries: List[QueryWithEmbedding]) -> List[QueryResult]:
+        """
+        Takes in a list of queries with embeddings and filters and
+        returns a list of query results with matching document chunks and scores.
+        """
+        # Prepare query responses and results object
+        results: List[QueryResult] = []
+
+        # Create a connection
+        conn = psycopg.connect(**self.conn_params)
+        cur = conn.cursor()
+
+        for query in queries:
+            query_results: List[DocumentChunkWithScore] = []
+            embedding = ",".join(str(v) for v in query.embedding)
+            query_statement = (
+                "SELECT doc_id, text, metadata, 1 - (%s.embedding <=> ARRAY[%s]::vector) AS score FROM %s ORDER BY score DESC LIMIT %d;"
```

Comment:
Does this hit the index with pgvector? I ask since it's `1 - (...)`. pgvector also has an index optimization for direct, that avoid some sqrt calls...

Reply:
I am using the following as reference from https://postgresml.org/blog/tuning-vector-recall-while-generating-query-embeddings-in-the-database. Is there a way to verify whether we are hitting the index?
```sql
WITH request AS (
  SELECT pgml.embed(
    'intfloat/e5-large',
    'Best 1980''s scifi movie'
  )::vector(1024) AS embedding
)

SELECT
  review_body,
  product_title,
  star_rating,
  total_votes,
  1 - (
    review_embedding_e5_large <=> (
      SELECT embedding FROM request
    )
  ) AS cosine_similiarity
FROM pgml.amazon_us_reviews
ORDER BY review_embedding_e5_large <=> (SELECT embedding FROM request)
LIMIT 5;
```

Reply:
This is slightly different, 

```ORDER BY review_embedding_e5_large <=> (SELECT embedding FROM request)``` 
is exactly a pgvector operator compatible with the IVFFLAT index. `1 - (...) <=> ...` is a formula that may be too complex for the planner to recognize, but I'm not sure.